src/compiler/checker.ts

```diff
@@ -4387,48 +4457,50 @@ namespace ts {
+                        return result;
+                    }
+                }
+                else if (relation !== identityRelation) {
+                    if (source.flags & TypeFlags.Union) {
+                        if (result = eachTypeRelatedToType(<UnionType>source, target, reportErrors)) {
+                            return result;
                         }
-                        else {
-                            if (result = unionTypeRelatedToType(<UnionType>target, source, reportErrors)) {
-                                return result;
-                            }
+                    }
+                    else if (target.flags & TypeFlags.Intersection) {
+                        if (result = typeRelatedToEachType(source, <IntersectionType>target, reportErrors)) {
+                            return result;
                         }
                     }
                     else {
-                        if (source.flags & TypeFlags.Union) {
-                            if (result = unionTypeRelatedToType(<UnionType>source, target, reportErrors)) {
+                        if (source.flags & TypeFlags.Intersection) {
+                            // If target is a union type the following check will report errors so we suppress them here
+                            if (result = someTypeRelatedToType(<IntersectionType>source, target, reportErrors && !(target.flags & TypeFlags.Union))) {
                                 return result;
                             }
                         }
-                        else {
-                            if (result = typeRelatedToUnionType(source, <UnionType>target, reportErrors)) {
+                        if (target.flags & TypeFlags.Union) {
+                            if (result = typeRelatedToSomeType(source, <UnionType>target, reportErrors)) {
                                 return result;
                             }
                         }
                     }
                 }
-                else if (source.flags & TypeFlags.TypeParameter && target.flags & TypeFlags.TypeParameter) {
-                    if (result = typeParameterRelatedTo(<TypeParameter>source, <TypeParameter>target, reportErrors)) {
-                        return result;
-                    }
-                }
-                else if (source.flags & TypeFlags.Reference && target.flags & TypeFlags.Reference && (<TypeReference>source).target === (<TypeReference>target).target) {
-                    // We have type references to same target type, see if relationship holds for all type arguments
-                    if (result = typesRelatedTo((<TypeReference>source).typeArguments, (<TypeReference>target).typeArguments, reportErrors)) {
-                        return result;
+                else {
+                    if (source.flags & TypeFlags.Union && target.flags & TypeFlags.Union ||

```

Comment:
I think it is easier to read when the identity case is on top, because it's hard to see what this `else` corresponds to.